filebeat/input/file.go

```diff
@@ -72,7 +77,37 @@ func mergeJSONFields(f *FileEvent, event common.MapStr) {
 	if f.JSONConfig.KeysUnderRoot {
 		for k, v := range f.JSONFields {
 			if f.JSONConfig.OverwriteKeys {
-				event[k] = v
+				if k == "@timestamp" {
+					vstr, ok := v.(string)
+					if !ok {
+						logp.Err("JSON: Won't overwrite @timestamp because value is not string")
+						event[jsonErrorKey] = "@timestamp not overwritten (not string)"
+						continue
+					}
+					// @timestamp must be of time common.Time
+					ts, err := common.ParseTime(vstr)
+					if err != nil {
+						logp.Err("JSON: Won't overwrite @timestamp because of parsing error: %v", err)
+						event[jsonErrorKey] = fmt.Sprintf("@timestamp not overwritten (parse error on %s)", vstr)
```

Comment:
I think it would be better to include the error and the value. Probably the error already includes the failed value.